### PR TITLE
[WIP] DNM Dgp to openflow hack

### PR DESCRIPTION
**Pull request by trozet**
Testing @numansiddique fix

### CODE REVIEW

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 1]
```diff
@@ -291,6 +368,45 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
...
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, table=4,ip,"+
+				"actions=ct(commit,zone=%d,nat(src=%s),table=3)",
+				defaultOpenFlowCookie, OVNMasqCTZone, types.V4OVNMasqueradeIP))
+	}
+	if config.IPv6Mode {
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, table=4,ip6, "+
+				"actions=ct(commit,zone=%d,nat(src=%s),table=3)",
+				defaultOpenFlowCookie, OVNMasqCTZone, types.V4OVNMasqueradeIP))
```
Comment (girishmg) [inline go-controller/pkg/node/gateway_shared_intf.go:394]: L394: types.V6OVNMasqueradeIP?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 2]
```diff
@@ -324,6 +440,51 @@ func newSharedGateway(nodeName string, subnets []*net.IPNet, gwNextHops []net.IP
...
+	bridgeLink, err := util.LinkSetUp(bridgeName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find shared gw bridge interface: %s", bridgeName)
+	}
+	for _, subnet := range config.Kubernetes.ServiceCIDRs {
+		nextHops, err := util.MatchIPFamily(utilnet.IsIPv6CIDR(subnet), gwNextHops)
+		if err != nil {
+			klog.Warningf("Unable to find valid next hop for Service CIDR: %s, available next hops: %v",
+				subnet, gwNextHops)
+		}
```
Comment (girishmg) [inline go-controller/pkg/node/gateway_shared_intf.go:453]: L453: should you continue here? as `nexthops` will be nil and you are accessing it at L456 below.

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 3]
```diff
@@ -324,6 +440,51 @@ func newSharedGateway(nodeName string, subnets []*net.IPNet, gwNextHops []net.IP
...
+
+		// TODO(trozet): see if we need to consider upgrade here...there may be an old route pointing to mp0
+		if exists, err := util.LinkRouteExists(bridgeLink, nextHops[0], subnet); err == nil && !exists {
```
Comment (girishmg) [inline go-controller/pkg/node/gateway_shared_intf.go:456]: L455: yes, I think we will need an upgrade path here.. unfortunately..

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:394]: ah thank you!

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:453]: yep I missed it. thanks. As you also mentioned to me before, we probably dont even need this service route. As long as we ensure the legacy route pointing to mp0 is deleted, that should be enough. I think the service subnet will then use the default gateway route which is all we need so that br-ex can take the packet and push it into OVN. Thoughts?

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 3 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:456]: ok will handle removing the old route.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:394]: also remind me to make a doc file to show how the packet flow will work to include in the final PR

File: go-controller/pkg/util/util.go [hunk 4]
```diff
@@ -188,3 +188,30 @@ func HashForOVN(s string) string {
 	return fmt.Sprintf("a%s", hashString)
 
 }
+
+// UpdateIPsSlice will search for values of v in the slice and update it with n values of same IP family
+func UpdateIPsSlice(s, oldIPs, newIPs []string) []string {
```
Comment (aojea) [inline go-controller/pkg/util/util.go:193]: I can't get from the description what is `v` and `n` :)

It looks that is easy to go wrong here and there can be corner cases like passing empty slices that can break
if you don't have time to add an unit test just add a TODO  to add it, is a good first  issue for someone new to the project ...

File: go-controller/pkg/types/const.go [hunk 5]
```diff
@@ -52,9 +52,26 @@ const (
...
+	V4OVNMasqueradeIP  = "169.254.169.1"
+	V6OVNMasqueradeIP  = "fd69::1"
+
 	// OpenFlow and Networking constants
 	RouteAdvertisementICMPType    = 134
 	NeighborAdvertisementICMPType = 136
 
 	OvnACLLoggingMeter = "acl-logging"
+
+	// LoadBalancer External Names
```
Comment (aojea) [inline go-controller/pkg/types/const.go:67]: it should be good to briefly explain what represent each category: cluster, worker, gateway

**Comment by trozet** [comment]
this passes ipv4 and ipv6. I did not add support for dual stack initially. I went back and re-added it and decided to make a new PR. @aojea PTAL and add your current comments there (sorry for the inconvience)

Moved to #2055
